{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, cbind, colMeans, colnames,\n",
      "    colSums, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, lengths, Map, mapply, match,\n",
      "    mget, order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    apply\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Attaching package: ‘scater’\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SingleCellExperiment)\n",
    "library(scater)\n",
    "options(stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "molecules <- read.table(\"tung/molecules.txt\", sep = \"\\t\")  # UMIs?\n",
    "anno <- read.table(\"tung/annotation.txt\", sep = \"\\t\", header=TRUE)\n",
    "# read_counts <- read.table(\"tung/reads.txt\", sep = \"\\t\")    \n",
    "#head(molecules[,1:3])\n",
    "head(anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## Standardise the analysis ##\n",
    "umi <- SingleCellExperiment(\n",
    "    assays = list(counts = as.matrix(molecules)),\n",
    "    colData = anno\n",
    ")\n",
    "\n",
    "# First, filter out genes not found in any cell\n",
    "genes_summed_gt_zero <- rowSums(counts(umi) > 0)\n",
    "gene_names_gt_zero <- (genes_summed_gt_zero > 0)\n",
    "\n",
    "umi_filt_genes <- umi[gene_names_gt_zero,]\n",
    "message(\"num Genes removed = \", dim(umi) - dim(umi_filt_genes))\n",
    "umi <- umi_filt_genes\n",
    "\n",
    "# Second, define control features (genes). \n",
    "# - ERCC spike-ins or mitochondrial genes\n",
    "# This can be defined either from specific gene names\n",
    "\n",
    "user_spike_filterwords = c(\"ERCC\")\n",
    "user_spike_samples = c(\n",
    "    \"ENSG00000198899\", \"ENSG00000198727\", \"ENSG00000198888\",\n",
    "    \"ENSG00000198886\", \"ENSG00000212907\", \"ENSG00000198786\",\n",
    "    \"ENSG00000198695\", \"ENSG00000198712\", \"ENSG00000198804\",\n",
    "    \"ENSG00000198763\", \"ENSG00000228253\", \"ENSG00000198938\", \"ENSG00000198840\")\n",
    "\n",
    "\n",
    "# populate a list of feature controls\n",
    "feat_conts <- list()\n",
    "\n",
    "for (u in 1:length(user_spike_filterwords)){\n",
    "    usk <- user_spike_filterwords[u]\n",
    "    usk_term <- sprintf(\"^%s-\", usk)\n",
    "    isSpike(umi, usk) <- grepl(usk_term, rownames(umi))\n",
    " \n",
    "    # append\n",
    "    vv <- c(feat_conts, isSpike(umi, usk))\n",
    "}\n",
    "\n",
    "isSpike(umi, \"MT\") <- rownames(umi) %in% user_spike_samples\n",
    "\n",
    "umi <- calculateQCMetrics(umi,       # feature_controls = feat_conts)\n",
    "    feature_controls = feat_conts\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell QC -- we need to decide the size of the library\n",
    "#  (how many molecules detected per sample).\n",
    "#\n",
    "# In the case of UMIs, this is the number of umi counts, as opposed to\n",
    "#  using the number of read counts.\n",
    "\n",
    "# Distribution of count numbers assosciated with a given UMI well\n",
    "# -- we can use this histogram to filter out wells with low read-depths\n",
    "#    and ensure sufficient \n",
    "\n",
    "user_thresh = 25000\n",
    "\n",
    "hist(umi$total_counts, breaks = 100)\n",
    "abline(v=user_thresh, col=\"red\")\n",
    "\n",
    "below_red <- sum(umi$total_counts < user_thresh)\n",
    "above_red <- sum(umi$total_counts >= user_thresh)\n",
    "\n",
    "message(\"below thresh = \", below_red, \", above thresh = \", above_red)\n",
    "# So we should remove 46 cells with low amount of read-depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can also look at the distribution of unique gene coverage\n",
    "#  in each sample\n",
    "\n",
    "# Here we are using high-depth coverage data, so we expect a high level of coverage\n",
    "#  for all detected genes in our set (this might NOT be that case for rare cell/droplet\n",
    "#  data).\n",
    "\n",
    "hist(umi$total_features, breaks = 100)\n",
    "# From this we can see that most cells have around ~ 7000-10000 genes\n",
    "#  which is normal for high-depth scRNA-seq.\n",
    "\n",
    "# we need to make a way to auto-detect this.\n",
    "autodetected_thresh = 7000\n",
    "abline(v = autodetected_thresh, col=\"red\")\n",
    "\n",
    "# CAUTION: This is only true for high-depth scRNA-seq. For lower-depth\n",
    "#          protocols such as droplet-based that are more tailed for\n",
    "#          rare cell, we would expect a lower distribution of reads.\n",
    "\n",
    "below_red <- sum(umi$total_features < autodetected_thresh)\n",
    "above_red <- sum(umi$total_features >= autodetected_thresh)\n",
    "\n",
    "message(\"below thresh = \", below_red, \", above thresh = \", above_red)\n",
    "# So we should remove 116 cells which have low range of genes detected\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotPhenoData(\n",
    "    umi,\n",
    "    aes_string(\n",
    "        x = \"total_features\",\n",
    "        y = \"pct_counts_MT\",\n",
    "        colour = \"batch\"\n",
    "    )\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
